File: whatsapp-worker/queue.go
package main

import (
	"context"
	"fmt"
	"time"

	"go.mau.fi/whatsmeow"
	"go.mau.fi/whatsmeow/proto/waE2E"
	"go.mau.fi/whatsmeow/types"
)

// Poll messages_a_envoyer (statut=en_attente), poste dans le groupe WhatsApp, marque comme
// envoyé. Diffusion à sens unique uniquement (voir plan-houville.md, section H) : aucune
// lecture de message entrant ici, juste l'envoi des messages générés par vercel-app.
func pollAndSend(client *whatsmeow.Client, sb *supabaseClient, groupJID string) error {
	jid, err := types.ParseJID(groupJID)
	if err != nil {
		return fmt.Errorf("WHATSAPP_GROUP_JID invalide : %w", err)
	}

	messages, err := sb.pendingMessages()
	if err != nil {
		return fmt.Errorf("lecture messages_a_envoyer : %w", err)
	}

	for _, m := range messages {
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		_, err := client.SendMessage(ctx, jid, &waE2E.Message{
			Conversation: &m.Contenu,
		})
		cancel()
		if err != nil {
			// Message laissé en_attente : sera retenté au prochain passage.
			fmt.Printf("Échec d'envoi du message %d : %v\n", m.ID, err)
			continue
		}
		if err := sb.markSent(m.ID); err != nil {
			fmt.Printf("Message %d envoyé mais marquage échoué : %v\n", m.ID, err)
		} else {
			fmt.Printf("Message %d envoyé et marqué.\n", m.ID)
		}
	}
	return nil
}

File: whatsapp-worker/src/queue.ts
import { getSocket, isWhatsAppConnected } from "./whatsapp.js";
import { supabase } from "./supabase.js";

// Poll messages_a_envoyer (statut=en_attente), poste dans le groupe WhatsApp, marque comme
// envoyé. Diffusion à sens unique uniquement (voir plan-houville.md, section H) : aucune
// lecture de message entrant ici, juste l'envoi des messages générés par vercel-app.

// Volontairement lu à chaque appel (pas figé au chargement du module) : au tout premier
// démarrage, on ne connaît pas encore le JID du groupe cible (il faut d'abord se connecter
// pour lister les groupes existants, voir scripts/list-groups.ts) — le worker doit pouvoir
// démarrer et se connecter sans planter avant que cette variable soit renseignée.
function getGroupJid(): string | undefined {
  return process.env.WHATSAPP_GROUP_JID;
}

// Timestamp du dernier passage de la boucle, quel que soit le résultat — utilisé par /health
// (section I du plan) pour détecter une boucle de polling bloquée.
let lastPollAt: Date | null = null;

export function getLastPollAt(): Date | null {
  return lastPollAt;
}

export async function pollAndSend(): Promise<void> {
  lastPollAt = new Date();

  if (!isWhatsAppConnected()) {
    console.log("Polling ignoré : WhatsApp non connecté pour l'instant.");
    return;
  }
  const groupJid = getGroupJid();
  if (!groupJid) {
    console.log("Polling ignoré : WHATSAPP_GROUP_JID pas encore configuré.");
    return;
  }
  const sock = getSocket();
  if (!sock) return;

  const { data: messages, error } = await supabase
    .from("messages_a_envoyer")
    .select("id, contenu")
    .eq("statut", "en_attente")
    .order("date_creation", { ascending: true });

  if (error) {
    console.error("Lecture messages_a_envoyer échouée :", error.message);
    return;
  }
  if (!messages || messages.length === 0) return;

  for (const message of messages) {
    try {
      await sock.sendMessage(groupJid, { text: message.contenu });

      const { error: errUpdate } = await supabase
        .from("messages_a_envoyer")
        .update({ statut: "envoye", date_envoi: new Date().toISOString() })
        .eq("id", message.id);
      if (errUpdate) {
        console.error(`Message ${message.id} envoyé mais marquage échoué :`, errUpdate.message);
      } else {
        console.log(`Message ${message.id} envoyé et marqué.`);
      }
    } catch (e) {
      // Message laissé en_attente : sera retenté au prochain passage.
      console.error(`Échec d'envoi du message ${message.id} :`, (e as Error).message);
    }
  }
}
